rootston/cursor.c

```diff
@@ -148,14 +154,37 @@ static void roots_passthrough_cursor(struct roots_cursor *cursor,
 	}
 }
 
+static void roots_grabbed_cursor(struct roots_cursor *cursor,
+		uint32_t time) {
+	struct roots_seat *seat = cursor->seat;
+	struct roots_view *view = cursor->grabbed_view;
+	double sx = cursor->cursor->x - view->x;
+	double sy = cursor->cursor->y - view->y;
+	if (view && view->wlr_surface) {
+		wlr_seat_pointer_notify_enter(seat->seat, view->wlr_surface, sx, sy);
+		wlr_seat_pointer_notify_motion(seat->seat, time, sx, sy);
+	}
+}
+
 static void roots_cursor_update_position(
 		struct roots_cursor *cursor, uint32_t time) {
 	struct roots_seat *seat = cursor->seat;
 	struct roots_view *view;
+
+	if (cursor->mode == ROOTS_CURSOR_PASSTHROUGH &&
+			seat->seat->pointer_state.button_count > 0 &&
+			cursor->grabbed_view &&
+			!wlr_seat_pointer_has_grab(seat->seat)) {
```

Comment:
(rewriting previous comment because github marked it as outdated before it went live)
I don't like changing cursor mode on move, if there is no drawback to entering ROOT_CURSOR_GRABBED with no button count I'd set it at click-within-a-grab time. When you set grabbed_view?

Reply:
This would be necessary for the behavior removed in 26c6c69, which is to deny popups (or other grabbing actions) when the pointer has a grab. This seems to be the behavior in X11 (try it) although not in Weston. Although in the X11 case, I'm not sure if this is enforced by the Xorg server or if clients voluntarily do not show popups in this case. The actual sending of buttons is the same in X11 and Weston as it is here however (you can see that clearly with urxvt).